File: EnumExampleC#/EnumExample/EnumUtils.cs
namespace EnumExample;

public sealed class EnumUtils
{
    private EnumUtils()
    {
    }

    public static T GetEnumByValue<T, V>(V value) where T : Enum, EnumWithValue<V> => Enum.Values<T>()
        .FirstOrDefault(x => x.Compare(value));

    public static T GetEnumByValueInList<T, V>(V value) where T : Enum, EnumWithValue<List<V>> => Enum.Values<T>()
        .FirstOrDefault(x => x.Value.Contains(value));

    public static T GetEnumByStringValue<T>(string value) where T : Enum, EnumWithValue => Enum.Values<T>()
        .FirstOrDefault(x => x.StringValue.Equals(value));

    public static T GetEnumByStringValueInList<T>(string value) where T : Enum, EnumWithValue => Enum.Values<T>()
        .FirstOrDefault(x => x.StringValue.Split(", ").Contains(value));
}

File: EnumExampleJava/src/main/java/gr/devian/enumexample/EnumUtils.java
package gr.devian.enumexample;

import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.function.Function;

public class EnumUtils {
    public static <T extends Enum<T> & EnumWithValue<V>, V> T getEnumByValue(final Class<T> enumClass, final V value) {
        return Arrays
            .stream(enumClass.getEnumConstants())
            .filter(x -> x.compare(value))
            .findFirst()
            .orElse(null);
    }

    public static <T extends Enum<T> & EnumWithValue<List<V>>, V> T getEnumByValueInList(final Class<T> enumClass, final V value) {
        return Arrays
            .stream(enumClass.getEnumConstants())
            .filter(x -> x.getValue().contains(value))
            .findFirst()
            .orElse(null);
    }

    public static <T extends Enum<T> & EnumWithValue<?>> T getEnumByStringValue(final Class<T> enumClass, final String value) {
        return Arrays
            .stream(enumClass.getEnumConstants())
            .filter(x -> Objects.equals(x.getStringValue().toLowerCase(), value.toLowerCase()))
            .findFirst()
            .orElse(null);
    }

    public static <T extends Enum<T> & EnumWithValue<List<?>>> T getEnumByStringValueInList(final Class<T> enumClass, final String value) {
        return Arrays
            .stream(enumClass.getEnumConstants())
            .filter(x -> Arrays.asList(x.getStringValue().split(", ")).contains(value))
            .findFirst()
            .orElse(null);
    }
}

